tests/IntSetValidity.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE CPP #-}
+module IntSetValidity (valid) where
+
+import Data.Bits (xor, (.&.))
+import Data.IntSet.Internal
+import Test.QuickCheck (Property, counterexample, property, (.&&.))
+
+{--------------------------------------------------------------------
+  Assertions
+--------------------------------------------------------------------}
+-- | Returns true iff the internal structure of the IntSet is valid.
+valid :: IntSet -> Property
+valid t =
+  counterexample "nilNeverChildOfBin" (nilNeverChildOfBin t) .&&.
+  counterexample "maskPowerOfTwo" (maskPowerOfTwo t) .&&.
+  counterexample "commonPrefix" (commonPrefix t) .&&.
+  counterexample "markRespected" (maskRespected t) .&&.
+  counterexample "tipsValid" (tipsValid t)
+
+-- Invariant: Nil is never found as a child of Bin.
+nilNeverChildOfBin :: IntSet -> Bool
+nilNeverChildOfBin t =
+  case t of
+    Nil -> True
+    Tip _ _ -> True
+    Bin _ _ l r -> noNilInSet l && noNilInSet r
+  where
+    noNilInSet t' =
+      case t' of
+        Nil -> False
+        Tip _ _ -> True
+        Bin _ _ l' r' -> noNilInSet l' && noNilInSet r'
+
+-- Invariant: The Mask is a power of 2.  It is the largest bit position at which
+--            two elements of the set differ.
+maskPowerOfTwo :: IntSet -> Bool
+maskPowerOfTwo t =
+  case t of
+    Nil -> True
+    Tip _ _ -> True
+    Bin _ m l r ->
+      (m `mod` 2 == 0) && maskPowerOfTwo l && maskPowerOfTwo r
```

Comment:
I am a bit late to the party, but

    m `mod` 2 == 0

is not a way to test  that a mask is a power of 2. Alternative: `popcount m == 1`.